Refactored and updated doclet, included documentation and sample project

manoelcampos opened the pull request:
- Performed some refactoring to better organize the code.
- Updated and included missing javadocs.
- Enabled the doclet to accept default command line arguments given to the javadoc tool (initially only the -d, that defines the output directory)
- Provided a complete README that describes the doclet, its goal and how to use it by directly calling the javadoc tool or through maven.
- Included a pre-configured sample maven project.

manoelcampos committed:
- Included missing copyright notice.
- Changed the javadoc style in the copyright notice to a normal java comment to avoid IDEs such as IntelliJ to issue a warning complaining about a "dangling javadoc" that is not linked to any element (such as a class, method, attribute, etc).

manoelcampos committed:
- Updated and included some missing javadocs

manoelcampos committed:
- Removed unused ExportDoclet.InnerClass

manoelcampos committed:
- Moved the code of the ExportDoclet.run() method to the start() because the start() method just delegates the call to the run() that is not called anywhere.

manoelcampos committed:
- Added misssing javadocs

manoelcampos committed:
- Renamed the instance method start() to render() and performed some refactoring

manoelcampos committed:
- Included README information and license badge.
- Small refactoring

manoelcampos committed:
- Updated the README including examples on how to use the doclet by directly calling the javadoc tool or configuring it inside a maven project.
- Included a sample project with everything configure.
- Updated the exportdoclet in order to make it accept javadoc command line arguments, such as the -d (that defines the output directory)

manoelcampos committed:
- Included the pom.xml snippets directly inside the README.adoc because GitHub interface doesn't parse include directives.

manoelcampos committed:
Updated README

manoelcampos committed:
- Included additional references

johncarl81 commented:
Fantastic @manoelcampos, ill review soon.

johncarl81 commented on README.adoc:
```diff
@@ -1,3 +1,139 @@
 
-image:https://travis-ci.org/johncarl81/exportdoclet.svg?branch=master["Build Status", link="https://travis-ci.org/johncarl81/exportdoclet"]
\ No newline at end of file
+image:https://travis-ci.org/johncarl81/exportdoclet.svg?branch=master["Build Status", link="https://travis-ci.org/johncarl81/exportdoclet"] image:http://img.shields.io/badge/license-ASF2-blue.svg["Apache License 2", link="http://www.apache.org/licenses/LICENSE-2.0.txt"]
```
Im a fan of the build status, but the license is shown now on the Github tab bar.  Remove?

manoelcampos replied:
Removed.

johncarl81 commented on README.adoc:
```diff
@@ -1,3 +1,139 @@
+ExportDoclet is a Java Maven project that can be built directly from any IDE or using the following maven command:
+
+[source,bash]
+mvn clean package install
```
I typically recommend just `mvn clean install` as `install` requires `package`

manoelcampos replied:
Done

johncarl81 commented on README.adoc:
```diff
@@ -1,3 +1,139 @@
+Now, to generate the javadocs in AsciiDoc files you can execute the `javadoc:javadoc` goal in Maven, using your IDE or the command line inside your project root directory:
+
+[source,bash]
+mvn javadoc:javadoc
```
It would be great to have gradle and ant examples as well (see [this example](https://github.com/asciidoctor/asciidoclet#usage))

manoelcampos replied:
Sorry, but I've never used Gradle and I don't know how to change the example so that it will work correctly. About ant, there are some years that I don't use it and I really don't know how anymore.

I'm wondering if you could help in this point or we could leave it to the community to improve.

johncarl81 replied:
Of course, lets leave this off for a future PR.

johncarl81 commented on README.adoc:
```diff
@@ -1,3 +1,139 @@
+
+=== Generating both the AsciiDoc and default HTML files
+
+If you want to generate both the AsciiDoc and HTML files, that requires two different configurations for the maven-javadoc-plugin, as exemplified below:
```
:+1: Nice, I didn't know about this technique!

johncarl81 commented on README.adoc:
```diff
@@ -1,3 +1,139 @@
+[source,bash]
+mvn javadoc:javadoc
+
+The exported javadocs to AsciiDoc files will be found at target/site/apidocs directory.
```
It would be great to show what to do with the output files, maybe with the asciidoctor-maven-plugin?

manoelcampos replied:
I extended the introduction and included the "How to use the generated AsciiDoc Files" section at the end.

johncarl81 commented on sample/pom.xml:
```diff
@@ -0,0 +1,125 @@
+
+    <groupId>org.asciidoctor</groupId>
+    <artifactId>exportdoclet-sample</artifactId>
+    <version>1.0</version>
```
I prefer the 1.0.0-SNAPSHOT (major.minor.bug) standard.

manoelcampos replied:
Yes, semantic versioning 😄. Done

johncarl81 commented on sample/pom.xml:
```diff
@@ -0,0 +1,125 @@
+                        <artifactId>exportdoclet</artifactId>
+                        <version>1.5.4-SNAPSHOT</version>
+                    </docletArtifact>
+                    <useStandardDocletOptions>true</useStandardDocletOptions>
```
Is this line necessary?

johncarl81 replied:
Ah: "Enabled the doclet to accept default command line arguments given to the javadoc tool (initially only the -d, that defines the output directory)"... nevermind.

manoelcampos replied:
OK

johncarl81 commented on sample/src/main/java/vehicles/AbstractVehicle.java:
```diff
@@ -0,0 +1,40 @@
+import java.time.LocalDate;
+
+/**
+ * An abstract class for implementing different kinds of {@link Vehicle}.
```
Why not use a Asciidoc link?

manoelcampos replied:
Particularly, I think the javadoc @link tag is nice, easy to write and it is extremely useful inside IDEs to navigate through the documentation, since asciidoc links are not parsed by IDEs.
But on the other hand, such javadoc tags will be exported as is to the adoc file and will not be parsed when generating the final documentation in PDF or any other format.

I think that avoiding the second issue may be better. However, I'm wondering if we could address the first issue and if you have some directions.

manoelcampos replied:
Apparently, any other javadoc tag such as `@author`, `@param`, `@return` will  not be parsed.
Such tags may be converted to asciidoc tags.
What I'm thinking it to make the exportdoclet to parse such javadoc tags and convert them to asciidoc ones.

johncarl81 replied:
Adding additional string processing is out of scope for this project.  We should stick to just exporting the javadoc strings and if additional processing is necessary it can be done after the fact.

manoelcampos replied:
Sorry to insist in this point, but I think this is a major issue. If such javadoc links are not converted to asciidoc ones, after the developer generates the final documentation for deployment, such elements will not be linked.

It is pretty easy do convert the link tag and it just requires very few lines of code. About the other tags such @author, @param and @return, maybe there is not need to convert. They would just be included after a "Parameters" section in order to simply organize the params, as it is already made in the default javadoc HTML.

johncarl81 replied:
If you feel strongly about this, please stick to your guns.  Discourse is one driving force behind high quality.  Could you implement this link conversion in a separate PR to keep this one focused?

manoelcampos replied:
For sure 👍

johncarl81 commented on sample/src/main/java/vehicles/Car.java:
```diff
@@ -0,0 +1,32 @@
+package vehicles;
+
+/**
+ * A Car represents a terrestrial {@link Vehicle}
```
Why not use a Asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportDoclet.java:
```diff
@@ -20,79 +20,80 @@
-
 /**
- * Class level comment
+ * A {@link Doclet} that exports javadoc comments containing AsciiDoc text to AsciiDoc files,
```
Why not use a Asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportDoclet.java:
```diff
@@ -20,79 +20,80 @@
     /**
-     * Private method comment
-     * @return
+     * Renders the javadoc documentation for all elements inside the {@link RootDoc} object
```
Same... Why not use a Asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportRenderer.java:
```diff
@@ -1,3 +1,18 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
FYI, you can update the licenses across the project by running the maven license plugin: `mvn license:format`

johncarl81 replied:
Which will format it starting witih `/**`

manoelcampos replied:
Good, in this case we can just configure the plugin to use /* instead.
Added the block below to the plugin configuration:
```xml
<mapping>
   <java>SLASHSTAR_STYLE</java>
</mapping>
```

johncarl81 replied:
Ah, nice.  Would you mind doing a `mvn license:format` and committing?

johncarl81 replied:
Does using the double-star style confuse the doclet and export the header?

manoelcampos replied:
Updated the copyright using the license plugin.
About the double-star, there is no issue in using it. It is just that some IDEs such as IntelliJ warns, since it isn't in fact a javadoc comment.

johncarl81 commented on src/main/java/org/asciidoctor/ExportRenderer.java:
```diff
@@ -11,28 +26,54 @@
 /**
+ * A renderer class that actually exports javadoc comments containing AsciiDoc text to AsciiDoc files,
+ * instead of specific final formats such as HTML.
+ * It is used when the {@link ExportDoclet} is started.
```
Same, asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportRenderer.java:
```diff
@@ -11,28 +26,54 @@
 
-    public boolean render(RootDoc rootDoc) {
+    /**
+     * Renders classes and packages javadocs, inside a {@link RootDoc} object, to AsciiDoc files.
```
Same, asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportRenderer.java:
```diff
@@ -79,24 +125,77 @@ public void renderPackage(PackageDoc doc){
     }
 
+    /**
+     * Exports a javadoc comment using a given {@link PrintWriter}, surrounding
```
Same, asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportRenderer.java:
```diff
@@ -79,24 +125,77 @@ public void renderPackage(PackageDoc doc){
+     *
+     * @param tag the name of the tag to surround the javadoc comment into the AsciiDoc file
+     * @param comment the javadoc comment to export
+     * @param writer the {@link PrintWriter} to be used to export the javadoc comment to an AsciiDoc file
```
Same, asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/ExportRenderer.java:
```diff
@@ -79,24 +125,77 @@ public void renderPackage(PackageDoc doc){
     }
 
+    /**
+     * Gets a {@link PrintWriter} to export the documentation of a class or package
```
Same, asciidoc link?

johncarl81 commented on src/main/java/org/asciidoctor/StandardAdapter.java:
```diff
@@ -16,11 +16,12 @@
 
 /**
- * Adapter class to use the Standard Javadoc Doclet in a non-static context.
+ * Adapter class to use the Standard Javadoc {@link Doclet} in a non-static context.
```
Same, asciidoc link?

johncarl81 commented:
Pretty good stuff.  Love the new example project.

johncarl81 commented on sample/src/main/java/vehicles/AbstractVehicle.java:
```diff
@@ -0,0 +1,40 @@
+ * @author http://twitter.com/manoelcampos[Manoel Campos da Silva Filho]
+ */
+public class AbstractVehicle implements Vehicle {
+    private int manufacturingYear;
```
Hmm, maybe show off a field javadoc export?

manoelcampos replied:
To avoid redundant documentation among the field and its getter and setter, I usually don't comment the field. Instead, I just use a @see tag to the getter (an asciidoc link in this case).
What do you think?

johncarl81 replied:
I completely agree.  I guess I'm just looking to demonstrate different aspect of the doclet.

manoelcampos replied:
Ok. Done.

manoelcampos commented:
I'm wondering if it's possible to used asciidoc include tag inside the README. 
I tried but it doesn't work and I haven't found any workaround.

manoelcampos commented:
Further, I don't know why it ignores the outputDirectory tag inside the pom.xml. Giving a different directory just works with the snippet below:

```xml
<additionalparam>-d "${project.build.directory}/site/asciidocs"</additionalparam>
```

manoelcampos committed:
- Updated README:
  ** Extending the introduction section and including a new "How to use the generated AsciiDoc Files" one.
  ** Removed the license badge.
  ** Simplified manve install command
  ** Included more details on how to use javadoc by command line.
- Changed license plugin to use the single star comment for the copyright notice
- Used a semantic version style for the sample project version number

johncarl81 commented:
I think Asciidoctor on github is limited, so imports may not work within a project.

manoelcampos committed:
- Changed javadoc link tag to asciidoc one.
- Added javadoc comments to fields.
- Included package documentation.

manoelcampos committed:
Added bintray and maven badges

manoelcampos committed:
Updated usage instructions

johncarl81 commented on README.adoc:
```diff
@@ -1,3 +1,156 @@
+
+Instead of exporting the javadoc comments to standard HTML files, the ExportDoclet exports them to AsciiDoc files, that in turn can be converted to any output format supported by the link:http://asciidoctor.org[AsciiDoctor toolchain].
+
+Using the ExportDoctor, one can put together the source code documentation (javadocs written in AsciiDoc), quick start guides, user guides, FAQs and any other documentation for Java projects. 
```
Change ExportDoctor to ExportDoclet.

manoelcampos replied:
Done

manoelcampos committed:
- Fix README typo and updated project description inside pom.xml
- Changed the build command from "mvn package" to "mvn clean package"

manoelcampos commented:
I think it's done. If there is anything else, please let me know.

johncarl81 commented:
Fantastic work @manoelcampos, thanks again!
